### PR TITLE
[WFLY-11651] added processor class for eval of method paramaters. Uni…

### PR DESCRIPTION
**Pull request by rsearls**
https://issues.jboss.org/browse/WFLY-11651
          [WFLY-11651] is dependent on https://issues.jboss.org/browse/WFCORE-4297
          [PR: https://github.com/wildfly/wildfly-core/pull/3655]
          WFCORE-4297 needs to be merged first so that this PR will compile successfully.
          WFCORE-4297 adds a constant to Phase.java that JaxrsSubsystemAdd references
          and associates with JaxrsMethodParameterProcessor.

### CODE REVIEW

**Review by marekkopecky — CHANGES_REQUESTED** [review]
I did some checks with this PR and https://github.com/wildfly/wildfly-core/pull/3655

* [Lazy annotation](https://github.com/jax-rs/api/blob/master/jaxrs-api/src/main/java/javax/ws/rs/ext/ParamConverter.java#L68-L71) should not be ignored. Lazy annotation is mentioned in the description of original [RESTEASY-2062](https://issues.jboss.org/browse/RESTEASY-2062) jira as well
* This PR stops deploying the application if DefaultValue annotation contains some wrong value. This approach is not synced with lazy behaviour (if Lazy annotation is used, http request is called, and resteasy finds wrong default value -> application is not undeployed, resteasy just prints error/exception). This could also caused some backward incompatibilities (although this is a bug fix, not a feature request). Probably better approach is to print error/warning and then deploy the deployment.

**Comment by rsearls** [comment]
Support added.

**Comment by marekkopecky** [comment]
Second part of my previous comment is still not addressed. It will also fix FormParamTest fail from RESTEasy TS, see the [jira comment](https://issues.jboss.org/browse/RESTEASY-2062?focusedCommentId=13687837&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13687837).

Btw, I don't see MediaTypeFromMessageBodyWriterTest error with this PR, https://github.com/wildfly/wildfly-core/pull/3655 and RESTEasy from 3.6 branch.

**Comment by marekkopecky** [comment]
With Lazy annotation:
* default value conversion should be deferred to a latest possible moment (i.e. until the injection of such default value is required)

Currently, it is evaluated during the deployment and warning (not error) is printed.

**Comment by rsearls** [comment]
I don't understand what you mean by, "Second part of my previous comment is 
still not addressed."  Per your Lazy annotation request,
"Probably better approach is to print error/warning and then deploy the deployment."
I have provided a warning message and allow the deployment to proceed when this
annotation is present.

**Comment by marekkopecky** [comment]
Basically:
* Lazy annotation is not use
   * I expect: startup warning + application is deployed
      * Because of backward compatibility and because spec allows this
   * Actual state with this PR: startup error + application is not deployed
* Lazy annotation is used
   * I expect: no startup warning  + application is deployed
      * Because DefaultValue annotation should not be evaluated with Lazy annotation during startup according to the spec
   * Actual state with this PR: startup warning  + application is deployed

**Comment by rsearls** [comment]
This PR is for work that is performed only in wfly during the "deployment"
phase.  The "startup" phase is performed in resteasy.  As we have discussed
previously, resteasy does not process an app's information until the initial 
call of a client, so to best address this under the current design constraints,
method parameters will be evaluated during the deployment phase in wfly and
parameter injection in resteasy which is a separate phase.  The time between the
warning provided by wfly and the error issued by resteasy will be variable.
The latter will depend in the initial client call to resteasy.

For the deployment phase in wfly I have changed the messages to be warnings only.
In resteasy the code originally provided for RESTEASY-2062 can be applied.
A PR for that will be provided at a later date.

**Comment by marekkopecky** [comment]
> This PR is for work that is performed only in wfly during the "deployment" phase. The "startup" phase is performed in resteasy.

Ok, let me say it according to these definitions:
* Lazy annotation is not use
   * I expect: deployment warning + application is deployed
   * Actual state with this PR: the same now :)
* Lazy annotation is used
   * I expect: no startup warning  + application is deployed
   * Actual state with this PR: the same as well

With this pure-wf approach, I guess that some additional RESTEasy changes are not necessary.

File: jaxrs/src/test/java/org/jboss/as/jaxrs/rsources/PrimitiveParamResource.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        return "content";
+    }
+
+    @GET
+    public String doGet(@QueryParam("float") @DefaultValue("3.14159265") float v) {
+        return "content";
+    }
+
+    @GET
+    //@Produces("application/double")
```
Comment (marekkopecky) [inline jaxrs/src/test/java/org/jboss/as/jaxrs/rsources/PrimitiveParamResource.java]: Can you remove these two comments?

File: jaxrs/src/main/java/org/jboss/as/jaxrs/logging/JaxrsLogger.java [hunk 2]
```diff
@@ -181,4 +181,21 @@
     @LogMessage(level = WARN)
     @Message(id = 18, value = "Explicit usage of Jackson annotation in a JAX-RS deployment; the system will disable JSON-B processing for the current deployment. Consider setting the '%s' property to 'false' to restore JSON-B.")
     void jacksonAnnotationDetected(String property);
```
[the remark this answers is not in the document]
Comment (rsearls) [reply jaxrs/src/main/java/org/jboss/as/jaxrs/logging/JaxrsLogger.java]: What do you want the message to say?  DummyParamConverterProvider provides an anonymous implementation of ParamConverter interface.  I am pointing the user to the specific class which is
the cause of the failure.

**Comment by bstansberry** [comment]
retest this please

File: jaxrs/src/main/java/org/jboss/as/jaxrs/logging/JaxrsLogger.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (marekkopecky) [reply jaxrs/src/main/java/org/jboss/as/jaxrs/logging/JaxrsLogger.java]: Debugging shows:
* `pc.getClass().getName()`
   * `org.resteasy.simple.deployment.DummyParamConverterProvider$1`
* `pc.toString()`
   * `org.resteasy.simple.deployment.DummyParamConverterProvider$1@21a0f094`

I'm looking to yasson (not related with this issue) exception example:
`07:35:57,999 SEVERE [org.eclipse.yasson.internal.Marshaller] (default task-1) Error getting value on: com.resteasy.test.Resource$1@6d78c412`

So it seems that $ (or even $@) format of anonymous class is quite common. @ronsigal WDYT?

**Comment by marekkopecky** [comment]
Btw,  once another commit will be added (or there appears decision that no other commit arrives), can you please rebase + force push to unblock team-city? There is currently this warning: "Cannot find commit 3501029b157e76a62d84dc790003c19cbc22b535", I'm not sure about the reason of this warning, but when I saw this last time, rebase + force push helps ...

File: jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsMethodParameterProcessor.java [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+                    detail.parameter.getName());
+
+            if (validators == null) {
+                // check for paramConverterProvider
+                validators = paramConverterMap.get(Object.class.getName());
+            }
+
+            boolean isCheckClazzMethods = true;
+            if (validators != null) {
+                for (Validator v : validators) {
```
Comment (marekkopecky) [inline jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsMethodParameterProcessor.java]: Optimization:
`try {...} catch { if (!lazy) log }` -> `if (!lazy) {try {...} catch(...) { log }}`

**Comment by rsearls** [comment]
Done.

**Comment by asoldano** [comment]
@rsearls , by looking at the failures in TC, I see two problems here:
1) JaxrsMethodParameterProcessorTestCase.fromValueTest failing because of NPE
2) some WS-Trust tests failing becase of a java.lang.NoClassDefFoundError: javax/ws/rs/core/Response in the code you're adding here (basically, it looks like a jaxws deployment is erroneously going through the jaxrs processor that you're adding)

**Comment by marekkopecky** [comment]
My previous comment is still not addressed: https://github.com/wildfly/wildfly/pull/12035#discussion_r255822264

**Comment by rsearls** [comment]
@marekkopecky I don't understand the nature of your complaint in comment #12035.

**Comment by bstansberry** [comment]
@rsearls @marekkopecky Did you come to an understanding since your last comments? I see the branch has been force pushed since then. But that may be unrelated.

If not I believe Marek's comment is stating he wants the fact that the problematic class is an anonymous inner class to be reflected in the message, by including the $1 part of the classname. Not including the $1 can be confusing as the outer class is not the problem.

**Comment by rsearls** [comment]
Thanks for that clarification,  that issue has been addressed in the latest push.  
Once TC testing is complete I believe this is ready for merging.

**Comment by bstansberry** [comment]
@asoldano Is this ok now?

**Comment by asoldano** [comment]
@bstansberry I think so, but maybe @marekkopecky wants to have the last word on this...?

**Comment by marekkopecky** [comment]
@asoldano All my previous comments are fixed now (thank you @rsearls and @bstansberry). I don't see any regression on upstream RESTEasy TS with this PR.

**Comment by bstansberry** [comment]
retest this please

**Pull request merged**